src/client/Microsoft.Identity.Client/Platforms/netdesktop/NetDesktopCryptographyManager.cs

```diff
@@ -85,10 +92,21 @@ public byte[] SignWithCertificate(string message, X509Certificate2 certificate)
                 return rsa.SignData(messageBytes, sha);
             }
 #else
-            using (var key = certificate.GetRSAPrivateKey())
+            if (!CertificateToRsaMap.TryGetValue(certificate.Thumbprint, out RSA rsa))
             {
-                return key.SignData(messageBytes, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);
+                if (CertificateToRsaMap.Count == maximumCertificateToRsaMap)
+                {
+                    ClearCertificateToRsaMap clearCertificateToRsaMap = new ClearCertificateToRsaMap(CertificateToRsaMap, clearMapLock);
+                    (new Thread(new ThreadStart(clearCertificateToRsaMap.Clear))).Start();
```

Comment:
If we really need a ConcurrentDictionary with a max count, it would be better to create a separate class for it, so that we can unit test it easily. I'd think that using `GetOrAdd`  and `Clear` might simplify some of this logic.

Reply:
ConcurrentDictionary.GetOrAdd could simplify, but need to be careful, overload that takes a valueFactory doesn't have a guarantee that the value returned from the function will be inserted into the dictionary.